src/option.rs

```diff
@@ -83,102 +83,102 @@ impl<T> From<Option<T>> for core::option::Option<T> {
-            inner.set_by_key(keys, de)
+        if let Some(inner) = self.0.as_ref() {
+            inner.get_by_key(keys, ser)
         } else {
             Err(Error::Absent(0))
         }
     }
 
-    fn get_by_key<K, S>(&self, keys: K, ser: S) -> Result<S::Error>
+    fn set_by_key<'a, K, D>(&mut self, keys: K, de: D) -> Result<usize, Error<D::Error>>
     where
         K: Iterator,
         K::Item: Key,
-        S: serde::Serializer,
+        D: serde::Deserializer<'a>,
     {
-        if let Some(inner) = self.0.as_ref() {
-            inner.get_by_key(keys, ser)
+        if let Some(inner) = self.0.as_mut() {
+            inner.set_by_key(keys, de)
         } else {
             Err(Error::Absent(0))
         }
     }
 
-    fn metadata() -> Metadata {
-        T::metadata()
-    }
-
-    fn traverse_by_key<K, F, E>(keys: K, func: F) -> Result<E>
+    fn traverse_by_key<K, F, E>(keys: K, func: F) -> Result<usize, Error<E>>
     where
         K: Iterator,
         K::Item: Key,
-        F: FnMut(Ok, usize, &str) -> core::result::Result<(), E>,
+        F: FnMut(usize, &str) -> Result<(), E>,
     {
         T::traverse_by_key(keys, func)
     }
+
+    fn metadata() -> Metadata {
+        T::metadata()
+    }
 }
 
 impl<T: serde::Serialize + serde::de::DeserializeOwned> Miniconf for core::option::Option<T> {
     fn name_to_index(_value: &str) -> core::option::Option<usize> {
         None
     }
 
-    fn set_by_key<'a, K, D>(&mut self, mut keys: K, de: D) -> Result<D::Error>
+    fn get_by_key<K, S>(&self, mut keys: K, ser: S) -> Result<usize, Error<S::Error>>
     where
         K: Iterator,
-        D: serde::Deserializer<'a>,
+        S: serde::Serializer,
     {
         if keys.next().is_some() {
             return Err(Error::TooLong(0));
         }
 
-        if let Some(inner) = self.as_mut() {
-            *inner = serde::Deserialize::deserialize(de)?;
-            Ok(Ok::Leaf(0))
+        if let Some(inner) = self.as_ref() {
+            serde::Serialize::serialize(inner, ser)?;
+            Ok(0)
         } else {
             Err(Error::Absent(0))
         }
     }
 
-    fn get_by_key<K, S>(&self, mut keys: K, ser: S) -> Result<S::Error>
+    fn set_by_key<'a, K, D>(&mut self, mut keys: K, de: D) -> Result<usize, Error<D::Error>>
     where
         K: Iterator,
-        S: serde::Serializer,
+        D: serde::Deserializer<'a>,
     {
         if keys.next().is_some() {
             return Err(Error::TooLong(0));
         }
 
-        if let Some(inner) = self.as_ref() {
-            serde::Serialize::serialize(inner, ser)?;
-            Ok(Ok::Leaf(0))
+        if let Some(inner) = self.as_mut() {
+            *inner = serde::Deserialize::deserialize(de)?;
+            Ok(0)
         } else {
             Err(Error::Absent(0))
         }
     }
 
+    fn traverse_by_key<K, F, E>(_keys: K, _func: F) -> Result<usize, Error<E>>
+    where
+        F: FnMut(usize, &str) -> Result<(), E>,
+    {
+        Ok(0)
```

Comment:
Why aren't we calling the function for the single element contained in the option?

Reply:
It was wrong to call it before. We never call it here if we don't add any path hierarchy levels here.